pydis_site/apps/api/viewsets/bot/off_topic_channel_name.py

```diff
@@ -106,7 +108,29 @@ def list(self, request: HttpRequest) -> Response:
                     'random_items': ["Must be a positive integer."]
                 })
 
-            queryset = self.get_queryset().order_by('?')[:random_count]
+            queryset = self.get_queryset().order_by('?').exclude(used=True)[:random_count]
+            self.get_queryset().filter(
+                name__in=(query.name for query in queryset)
+            ).update(used=True)
+
+            # When the client requests more channel names than are available,
+            # we reset all names to used=False and start a new round of names.
+            if len(queryset) < random_count:
+                # Figure out how many additional names we need, and don't fetch duplicate names.
+                names_needed = random_count - len(queryset)
+                other_names = self.get_queryset().order_by('?').exclude(
+                    name__in=(query.name for query in queryset)
+                )[:names_needed]
+
+                # Reset the `used` field to False for all names except the ones we just used.
+                self.get_queryset().exclude(name__in=(
+                    query.name for query in other_names)
+                ).update(used=False)
+
+                # Join original queryset (that had missing names)
+                # and extension with these missing names.
+                queryset = list(queryset) + list(other_names)
```

Comment:
I really had a feeling there must be a simpler way to do it. I played around with a bunch of ideas and ultimately came up with this:

```py
queryset = self.get_queryset().order_by('used', '?')[:random_count]

if any(obj.used for obj in queryset):
    self.get_queryset().update(
        used=Case(
            When(name__in=(obj.name for obj in queryset), then=Value(True)),
            default=Value(False)
        )
    )
else:
    self.get_queryset().filter(
        name__in=(obj.name for obj in queryset)
    ).update(used=True)
```

It's shorter and has only two queries rather than four. However, I don't know if it's strictly better. Maybe it's harder to follow. I'm not sure about performance since it does extra sorting and has an `any()` which iterates results. However, I imagine this to be negligible in practice given the amount of names we request. Furthermore, it is likely made up by the removal of the extra operations compared to the previous approach.

Let me know what you think. @lemonsaurus, maybe you have some comments since you reviewed it too?

Reply:
I welcome any simplification - I shared your gut feeling that there had to be a simpler way, but didn't have a great suggestion.

I think your approach looks smart, but it would require a few comments to not be scary. Especially the if clause.